source/portable/NetworkInterface/MPS3_AN552/NetworkInterface.c

```diff
@@ -0,0 +1,367 @@
+        xEthernetInterface0.pxEthernetMACDriver->PHY_Write );
+    xEthernetInterface0.pxEthernetPHYDriver->PowerControl( ARM_POWER_FULL );
+    xEthernetInterface0.pxEthernetPHYDriver->SetInterface( xEthernetMACCapabilities.media_interface );
+    xEthernetInterface0.pxEthernetPHYDriver->SetMode( ARM_ETH_PHY_AUTO_NEGOTIATE );
+
+    /* Wait for the Ethernet link to be up */
+    while( eEthernetLinkState == ARM_ETH_LINK_DOWN )
+    {
+        vTaskDelay( pdMS_TO_TICKS( 1000 ) );
+        eEthernetLinkState = xEthernetInterface0.pxEthernetPHYDriver->GetLinkState();
+        eEthernetLinkStateTimeOutCount++;
+
+        if( eEthernetLinkStateTimeOutCount >= 5 )
+        {
+            FreeRTOS_debug_printf( ( "Ethernet Link state is DOWN \n " ) );
+            xReturn = pdFAIL;
+        }
+    }
+
+    if( xReturn == pdPASS )
+    {
+        xEthernetLinkInfo = xEthernetInterface0.pxEthernetPHYDriver->GetLinkInfo();
+        ulMACControlArg = xEthernetLinkInfo.speed << ARM_ETH_MAC_SPEED_Pos |
+                          xEthernetLinkInfo.duplex << ARM_ETH_MAC_DUPLEX_Pos |
+                          ARM_ETH_MAC_ADDRESS_BROADCAST;
+        /* Configure Ethernet MAC based on PHY status */
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONFIGURE, ulMACControlArg );
+        /* Enable RX/TX */
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONTROL_RX, 1 );
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONTROL_TX, 1 );
+    }
+
+    return xReturn;
+}
+
+static void prvEthernetDriverNotifications( uint32_t event )
+{
+    BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+    configASSERT( xRxTaskHandle );
+
+    if( event == ARM_ETH_MAC_EVENT_RX_FRAME )
+    {
+        /* Ethernet frame received. Send notification to Receive task. */
+        vTaskNotifyGiveFromISR( xRxTaskHandle, &xHigherPriorityTaskWoken );
+    }
+
+    portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
+}
+/*-----------------------------------------------------------*/
+
+static BaseType_t xLAN91C111_NetworkInterfaceOutput( NetworkInterface_t * pxInterface,
+                                                     NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                                     BaseType_t xReleaseAfterSend )
+{
+    BaseType_t xReturn = pdFAIL;
+
+    ( void ) pxInterface;
+
+    if( pxNetworkBuffer->xDataLength > 1518 )
+    {
+        FreeRTOS_debug_printf( ( "Error: Too big Ethernet frame, Size:", pxNetworkBuffer->xDataLength ) );
+        return xReturn;
+    }
+
+    xReturn = xEthernetInterface0.pxEthernetMACDriver->SendFrame(
+        pxNetworkBuffer->pucEthernetBuffer,
+        pxNetworkBuffer->xDataLength,
+        0 );
+
+    if( xReturn < 0 )
+    {
+        FreeRTOS_debug_printf( ( "Error: Ethernet driver SendFrame returned %d ", xReturn ) );
+    }
+
+    if( xReleaseAfterSend == pdTRUE )
+    {
+        vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
```

Comment:
What about zero-copy? Does `SendFrame()` wait until the entire frame has been queued / sent?
If not the space will be released and may get corrupted.

Reply:
The driver doesn't support zero-copy. I've added a compile time error to catch this.